lib/Parse/Lexer.cpp

```diff
@@ -1887,11 +1887,17 @@ void Lexer::lexStringLiteral(unsigned CustomDelimiterLen) {
       } else if ((*CurPtr == '\r' || *CurPtr == '\n') && IsMultilineString) {
         // The only case we reach here is unterminated single line string in the
         // interpolation. For better recovery, go on after emitting an error.
-        diagnose(CurPtr, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
         wasErroneous = true;
         continue;
       } else {
-        diagnose(TokStart, diag::lex_unterminated_string);
+        diagnose(CurPtr, diag::string_interpolation_unclosed);
+        char *fixItLoc = CurPtr - 1;
+        if (*fixItLoc == QuoteChar) {
+          fixItLoc -= CustomDelimiterLen;
+        }
+        diagnose(fixItLoc, diag::string_interpolation_unclosed_fix_it)
+            .fixItInsert(Lexer::getSourceLoc(fixItLoc), ")");
         return formToken(tok::unknown, TokStart);
       }
```

Comment:
Entering this branch could really mean anything; a missing closing paren is just one possibility. If we are lexing a multiline string, interpolations are also allowed to span multiple lines, which means that we cannot provide a reasonable fix-it, nor can we be sure that it failed because of a missing paren.

With single-line strings, it's probably safe to assume that there is a missing paren *if* `CurPtr` ends up pointing to a line break or the end of the file, but it appears we still have no way of figuring out the fix-it location in general. Consider
```swift
// The intention is \(x)
//                         CurPtr
//                         ↓
let s = "printed \(x lines"
```

If `CurPtr` points to something else, however, our only choice without making this change a lot more involved is to emit a fallback "unterminated string" error — the initial purpose of this branch.

Reply:
Hm, should we just check for if it’s multiline with the already existing variable in the function, and if it’s not, emit the unclosed interpolation error, otherwise emit the unterminated string literal error?

Reply:
Sounds good. I think we can use the previous branch for that to keep the flow simple, i.e.

```c++
} else if (*CurPtr == '\r' || *CurPtr == '\n') {
  if (IsMultilineString) ... else ...
} else { 
  // unterminated string literal error
}
```

Reply:
Fantastic! I’ll do that, give me a second